### PR TITLE
`getReferencesAtLocation`: fix handling of destructoring imports

### LINKED ISSUE #45423: Cannot read property 'exportSymbol' of undefined
(Migrated from microsoft/tsserverfuzzer#315)

**Repo Name**: https://github.com/prettier/prettier.git

**Revision**: 1b2c72df101208310bc22964ac06766a5453f5d5

**TypeScript Version**: 4.3.0-dev.20210415

**Commands**: 

```
{"response":{"canRename":false,"localizedErrorMessage":"You cannot rename this element."},"responseRequired":true}
{"command":"rename-full","arguments":{"position":219,"file":"D:/local/Temp/fuzzerWorkingDir/prettier/src/document/index.js"},"type":"request","seq":25}
{"response":{"canRename":false,"localizedErrorMessage":"You cannot rename this element."},"responseRequired":true}
{"command":"rename-full","arguments":{"position":31,"file":"D:/local/Temp/fuzzerWorkingDir/prettier/src/document/index.js"},"type":"request","seq":22}
{"response":{"canRename":false,"localizedErrorMessage":"You cannot rename this element."},"responseRequired":true}
{"command":"rename-full","arguments":{"position":203,"file":"D:/local/Temp/fuzzerWorkingDir/prettier/src/document/index.js"},"type":"request","seq":24}
{"response":{"canRename":true,"kind":"property","displayName":"debug","fullDisplayName":"__object.debug","kindModifiers":"","triggerSpan":{"start":201,"length":5}},"responseRequired":true}
{"command":"rename-full","arguments":{"position":203,"file":"D:/local/Temp/fuzzerWorkingDir/prettier/src/document/index.js"},"type":"request","seq":25}
{"response":{"canRename":true,"kind":"property","displayName":"debug","fullDisplayName":"__object.debug","kindModifiers":"","triggerSpan":{"start":201,"length":5}},"responseRequired":true}
{"command":"renameLocations-full","arguments":{"position":203,"file":"D:/local/Temp/fuzzerWorkingDir/prettier/src/document/index.js"},"type":"request","seq":26}
(Showing last 10 of 35733 commands)
```

**Code** ([`src\document\index.js`](https://github.com/prettier/prettier/blob/1b2c72df101208310bc22964ac06766a5453f5d5/src/document/index.js)):

```js
    printer: require("./doc-printer"),
    utils: require("./doc-utils"),
    debug: require("./doc-debug"),
//    ^
  };
  
```

**Hash**: 3f8acbaf45f4d44239e3f4a341f4b8af
**Stack trace**: 
```
Cannot read property 'exportSymbol' of undefined
    at getExport (tsserverlibrary.js:127386:28)
    at getImportOrExportSymbol (tsserverlibrary.js:127381:53)
    at getImportOrExportReferences (tsserverlibrary.js:128991:56)
    at getReferencesAtLocation (tsserverlibrary.js:128921:17)
    at getReferencesInContainer (tsserverlibrary.js:128861:21)
    at getReferencesInSourceFile (tsserverlibrary.js:128848:24)
    at searchForName (tsserverlibrary.js:128642:21)
    at getReferencesInContainerOrFiles (tsserverlibrary.js:128413:25)
    at getReferencedSymbolsForSymbol (tsserverlibrary.js:128397:21)
    at getReferencedSymbolsForNode (tsserverlibrary.js:128163:34)
```

### PR DESCRIPTION
**Pull request by elibarzilay**
Fixes #45423.

### CODE REVIEW

File: src/services/findAllReferences.ts [hunk 1]
```diff
@@ -1523,9 +1523,14 @@ namespace ts.FindAllReferences {
...
 
             // Use the parent symbol if the location is commonjs require syntax on javascript files only.
-            referenceSymbol = isInJSFile(referenceLocation) && referenceLocation.parent.kind === SyntaxKind.BindingElement && isRequireVariableDeclaration(referenceLocation.parent)
-                ? referenceLocation.parent.symbol
-                : referenceSymbol;
+            if (isInJSFile(referenceLocation)
+                && referenceLocation.parent.kind === SyntaxKind.BindingElement
+                && isRequireVariableDeclaration(referenceLocation.parent)) {
+                referenceSymbol = referenceLocation.parent.symbol;
+                // The parent will not have a symbol if it's an ObjectBindingPattern (when destruturing is used).  In
```
Comment (RyanCavanaugh) [inline src/services/findAllReferences.ts:1530]: Please fix the typo (destruturing) so we don't get a PR from someone else later 🙂

**Comment by RyanCavanaugh** [comment]
GitHub is refusing to show `tests/cases/fourslash/findAllReferencesJsRequireDestructuring.ts` - can you post it as text (it's allegedly 13 lines)?

File: src/services/findAllReferences.ts (hunk 1 above)
Comment (elibarzilay) [reply src/services/findAllReferences.ts]: Bah!  Sorry -- fixed.  (Dusty keyboard...)

**Comment by elibarzilay** [comment]
@RyanCavanaugh, there was no change to `findAllReferencesJsRequireDestructuring.ts` -- I worked on it on the way to the real test and my Emacs adds the required-by-convention newline at the end of the file (so it was a whitespace change which GH doesn't handle well).  I reverted that line so it's gone from the pr now.

**Review by mkg652 — CHANGES_REQUESTED** [review]
J

**Comment by elibarzilay** [comment]
ping @armanio123 @andrewbranch

**Review by andrewbranch — APPROVED** [review]

**Pull request merged**